driver-core/src/main/com/mongodb/operation/QueryBatchCursor.java

```diff
@@ -108,10 +108,9 @@
         }
     }
 
-    private boolean canConnectionBeExhaust() {
-        Connection connection = connectionSource.getConnection();
+    private boolean canConnectionBeExhaust(final Connection connection) {
         if (serverIsAtLeastVersionFourDotOne(connection.getDescription())) {
-            exhaustConnection = connection;
+            exhaustConnection = connection.retain();
```

Comment:
One nit.  Set this in the constructor instead.  Methods that return booleans like this shouldn't have side effects.

Reply:
Done